llvm/include/llvm/ADT/SmallVector.h

```diff
@@ -99,8 +99,9 @@ template <class Size_T> class SmallVectorBase {
   ///
   /// This does not construct or destroy any elements in the vector.
   void set_size(size_t N) {
-    assert(N <= capacity());
-    Size = N;
+    auto n = static_cast<Size_T>(N);
+    assert(n <= capacity());
```

Comment:
would it make sense to also assert that `n == N`?

Reply:
`capacity()` also returns `size_t`, so probably better to keep the assertion first. e.g.

```
    assert(N <= capacity());
    Size = static_cast<Size_T>(N);

    // As Felipe suggests, you could add:
    assert(static_cast<size_t>(Size) == N);
```

Reply:
IMO, @browneee 's suggestion here is more clear / easier to follow. `n` vs. `N` is pretty subtle.

Reply:
Thinking a little more: `N <= capacity()` implies  `static_cast<size_t>(Size) == N` because `capacity()` is from a Size_T variable.

Reply:
> capacity() also returns size_t, so probably better to keep the assertion first.

good point. I'm not sure why our downstream had it differently. "reverted" this part back.

overall, rewrote this part as per @browneee 's advice. I think it makes sense to keep the `assert(static_cast<size_t>(Size) == N);` at the end as well. if anything, it's a function's postcondition validation so should be no harm (also, wondering if it might actually catch any data-losing `size_t -> uint32_t` narrowings).

Reply:
> I think it makes sense to keep the assert(static_cast<size_t>(Size) == N); at the end as well.

This assertion is correct, but not necessary because it is implied by `assert(N <= capacity());`.

I would avoid an unnecessary assertion, because this is performance sensitive code.

If you want to keep it for clarity then I'd suggest only doing it in debug:
```
LLVM_DEBUG(assert(static_cast<size_t>(Size) == N));
```

My vote would be to omit it, because it is covered by the earlier assertion.

Reply:
I thought assert() is by definition debug-only (exception being if one explicitly sets `LLVM_ENABLE_ASSERTIONS`). do you think people  use `assert()` via LLVM_ENABLE_ASSERTIONS in production?

Reply:
Since it's covered by the earlier assertion, seems like all we need is a comment explaining that. Anyone sufficiently interested can read the comment.

Reply:
added the comment.